[BASE-20][BASE-44] fixes for thrift client metrics

krav opened the pull request:
<!-- If this pull request closes an issue, please mention the issue number below -->
Closes [#BASE-20](https://reddit.atlassian.net/browse/BASE-20)
Closes [#BASE-44](https://reddit.atlassian.net/browse/BASE-44)


## 💸 TL;DR

* set protocol on thrift client

fixing observer ignoring thrift client metrics

* ensure span is started before handling request

when an exception was encountered before the span was started (e.g. remote thrift server is down), an
additional exception was raised by the `_call_thrift_method` exception handler, as it expected the span to have been started.

this changes the semantics of the latency graph, as it now includes every retry of a request!


## 📜 Details

Example error w stacktrace:

```{"message": "Unexpected exception in handler", "funcName": "process_example", "lineno": 163, "module": "AltTextService", "name": "root", "pathname": "/src/alt_text/alt_text_thrift/AltTextService.py", "process": 7, "processName": "MainProcess", "thread": 140516595254944, "threadName": "7037294429657837072", "exc_info": "Traceback (most recent call last):\n  File \"/usr/local/lib/venv/lib/python3.9/site-packages/baseplate/clients/thrift.py\", line 172, in _call_thrift_method\n    with self.pool.connection() as prot:\n  File \"/usr/local/lib/python3.9/contextlib.py\", line 119, in __enter__\n    return next(self.gen)\n  File \"/usr/local/lib/venv/lib/python3.9/site-packages/baseplate/lib/thrift_pool.py\", line 237, in connection\n    prot = self._create_connection()\n  File \"/usr/local/lib/venv/lib/python3.9/site-packages/baseplate/lib/thrift_pool.py\", line 202, in _create_connection\n    raise TTransportException(\nthrift.transport.TTransport.TTransportException: giving up after multiple attempts to connect\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/src/alt_text/alt_text_thrift/AltTextService.py\", line 154, in process_example\n    result.success = self._handler.example()\n  File \"/usr/local/lib/venv/lib/python3.9/site-packages/baseplate/frameworks/thrift/__init__.py\", line 48, in call_with_context\n    result = handler_fn(self.context, *args, **kwargs)\n  File \"/src/alt_text/handlers.py\", line 37, in example\n    result = are_strings_good(context, [\"something something\"])\n  File \"/src/alt_text/goodcontent.py\", line 28, in are_strings_good\n    response = goodcontent_svc.are_strings_good(request)\n  File \"/usr/local/lib/venv/lib/python3.9/site-packages/baseplate/clients/thrift.py\", line 218, in _call_thrift_method\n    span.finish(exc_info=sys.exc_info())\nUnboundLocalError: local variable 'span' referenced before assignment", "level": "ERROR", "traceID": "7037294429657837072"}```

Example output:
```
# HELP thrift_client_latency_seconds Multiprocess metric
# TYPE thrift_client_latency_seconds histogram
thrift_client_latency_seconds_sum{thrift_slug="goodcontent",thrift_success="true"} 0.005589752999999999
thrift_client_latency_seconds_bucket{le="0.0001",thrift_slug="goodcontent",thrift_success="true"} 0.0
thrift_client_latency_seconds_bucket{le="0.00025",thrift_slug="goodcontent",thrift_success="true"} 0.0
thrift_client_latency_seconds_bucket{le="0.000625",thrift_slug="goodcontent",thrift_success="true"} 0.0
thrift_client_latency_seconds_bucket{le="0.0015625",thrift_slug="goodcontent",thrift_success="true"} 0.0
thrift_client_latency_seconds_bucket{le="0.00390625",thrift_slug="goodcontent",thrift_success="true"} 3.0
thrift_client_latency_seconds_bucket{le="0.009765625",thrift_slug="goodcontent",thrift_success="true"} 3.0
thrift_client_latency_seconds_bucket{le="0.0244140625",thrift_slug="goodcontent",thrift_success="true"} 3.0
thrift_client_latency_seconds_bucket{le="0.06103515625",thrift_slug="goodcontent",thrift_success="true"} 3.0
thrift_client_latency_seconds_bucket{le="0.152587890625",thrift_slug="goodcontent",thrift_success="true"} 3.0
thrift_client_latency_seconds_bucket{le="0.3814697265625",thrift_slug="goodcontent",thrift_success="true"} 3.0
thrift_client_latency_seconds_bucket{le="0.95367431640625",thrift_slug="goodcontent",thrift_success="true"} 3.0
thrift_client_latency_seconds_bucket{le="2.384185791015625",thrift_slug="goodcontent",thrift_success="true"} 3.0
thrift_client_latency_seconds_bucket{le="5.9604644775390625",thrift_slug="goodcontent",thrift_success="true"} 3.0
thrift_client_latency_seconds_bucket{le="14.901161193847656",thrift_slug="goodcontent",thrift_success="true"} 3.0
thrift_client_latency_seconds_bucket{le="+Inf",thrift_slug="goodcontent",thrift_success="true"} 3.0
thrift_client_latency_seconds_count{thrift_slug="goodcontent",thrift_success="true"} 3.0
# HELP thrift_server_latency_seconds Multiprocess metric
# TYPE thrift_server_latency_seconds histogram
thrift_server_latency_seconds_sum{thrift_method="example",thrift_success="true"} 0.090386687
thrift_server_latency_seconds_sum{thrift_method="is_healthy",thrift_success="true"} 0.013437850000000003
thrift_server_latency_seconds_sum{thrift_method="example",thrift_success="false"} 2.094433138
thrift_server_latency_seconds_bucket{le="0.0001",thrift_method="example",thrift_success="true"} 0.0
thrift_server_latency_seconds_bucket{le="0.00025",thrift_method="example",thrift_success="true"} 0.0
thrift_server_latency_seconds_bucket{le="0.000625",thrift_method="example",thrift_success="true"} 0.0
thrift_server_latency_seconds_bucket{le="0.0015625",thrift_method="example",thrift_success="true"} 0.0
thrift_server_latency_seconds_bucket{le="0.00390625",thrift_method="example",thrift_success="true"} 0.0
thrift_server_latency_seconds_bucket{le="0.009765625",thrift_method="example",thrift_success="true"} 2.0
thrift_server_latency_seconds_bucket{le="0.0244140625",thrift_method="example",thrift_success="true"} 2.0
thrift_server_latency_seconds_bucket{le="0.06103515625",thrift_method="example",thrift_success="true"} 2.0
thrift_server_latency_seconds_bucket{le="0.152587890625",thrift_method="example",thrift_success="true"} 3.0
thrift_server_latency_seconds_bucket{le="0.3814697265625",thrift_method="example",thrift_success="true"} 3.0
thrift_server_latency_seconds_bucket{le="0.95367431640625",thrift_method="example",thrift_success="true"} 3.0
thrift_server_latency_seconds_bucket{le="2.384185791015625",thrift_method="example",thrift_success="true"} 3.0
thrift_server_latency_seconds_bucket{le="5.9604644775390625",thrift_method="example",thrift_success="true"} 3.0
thrift_server_latency_seconds_bucket{le="14.901161193847656",thrift_method="example",thrift_success="true"} 3.0
thrift_server_latency_seconds_bucket{le="+Inf",thrift_method="example",thrift_success="true"} 3.0
thrift_server_latency_seconds_count{thrift_method="example",thrift_success="true"} 3.0
thrift_server_latency_seconds_bucket{le="0.0001",thrift_method="is_healthy",thrift_success="true"} 7.0
thrift_server_latency_seconds_bucket{le="0.00025",thrift_method="is_healthy",thrift_success="true"} 79.0
thrift_server_latency_seconds_bucket{le="0.000625",thrift_method="is_healthy",thrift_success="true"} 82.0
thrift_server_latency_seconds_bucket{le="0.0015625",thrift_method="is_healthy",thrift_success="true"} 82.0
thrift_server_latency_seconds_bucket{le="0.00390625",thrift_method="is_healthy",thrift_success="true"} 82.0
thrift_server_latency_seconds_bucket{le="0.009765625",thrift_method="is_healthy",thrift_success="true"} 82.0
thrift_server_latency_seconds_bucket{le="0.0244140625",thrift_method="is_healthy",thrift_success="true"} 82.0
thrift_server_latency_seconds_bucket{le="0.06103515625",thrift_method="is_healthy",thrift_success="true"} 82.0
thrift_server_latency_seconds_bucket{le="0.152587890625",thrift_method="is_healthy",thrift_success="true"} 82.0
thrift_server_latency_seconds_bucket{le="0.3814697265625",thrift_method="is_healthy",thrift_success="true"} 82.0
thrift_server_latency_seconds_bucket{le="0.95367431640625",thrift_method="is_healthy",thrift_success="true"} 82.0
thrift_server_latency_seconds_bucket{le="2.384185791015625",thrift_method="is_healthy",thrift_success="true"} 82.0
thrift_server_latency_seconds_bucket{le="5.9604644775390625",thrift_method="is_healthy",thrift_success="true"} 82.0
thrift_server_latency_seconds_bucket{le="14.901161193847656",thrift_method="is_healthy",thrift_success="true"} 82.0
thrift_server_latency_seconds_bucket{le="+Inf",thrift_method="is_healthy",thrift_success="true"} 82.0
thrift_server_latency_seconds_count{thrift_method="is_healthy",thrift_success="true"} 82.0
thrift_server_latency_seconds_bucket{le="0.0001",thrift_method="example",thrift_success="false"} 0.0
thrift_server_latency_seconds_bucket{le="0.00025",thrift_method="example",thrift_success="false"} 0.0
thrift_server_latency_seconds_bucket{le="0.000625",thrift_method="example",thrift_success="false"} 0.0
thrift_server_latency_seconds_bucket{le="0.0015625",thrift_method="example",thrift_success="false"} 0.0
thrift_server_latency_seconds_bucket{le="0.00390625",thrift_method="example",thrift_success="false"} 0.0
thrift_server_latency_seconds_bucket{le="0.009765625",thrift_method="example",thrift_success="false"} 0.0
thrift_server_latency_seconds_bucket{le="0.0244140625",thrift_method="example",thrift_success="false"} 0.0
thrift_server_latency_seconds_bucket{le="0.06103515625",thrift_method="example",thrift_success="false"} 1.0
thrift_server_latency_seconds_bucket{le="0.152587890625",thrift_method="example",thrift_success="false"} 2.0
thrift_server_latency_seconds_bucket{le="0.3814697265625",thrift_method="example",thrift_success="false"} 2.0
thrift_server_latency_seconds_bucket{le="0.95367431640625",thrift_method="example",thrift_success="false"} 2.0
thrift_server_latency_seconds_bucket{le="2.384185791015625",thrift_method="example",thrift_success="false"} 4.0
thrift_server_latency_seconds_bucket{le="5.9604644775390625",thrift_method="example",thrift_success="false"} 4.0
thrift_server_latency_seconds_bucket{le="14.901161193847656",thrift_method="example",thrift_success="false"} 4.0
thrift_server_latency_seconds_bucket{le="+Inf",thrift_method="example",thrift_success="false"} 4.0
thrift_server_latency_seconds_count{thrift_method="example",thrift_success="false"} 4.0
# HELP bp_thrift_pool_max_size Multiprocess metric
# TYPE bp_thrift_pool_max_size gauge
bp_thrift_pool_max_size{client_cls="type",pid="7"} 0.0
# HELP bp_thrift_pool_active_connections Multiprocess metric
# TYPE bp_thrift_pool_active_connections gauge
bp_thrift_pool_active_connections{client_cls="type",pid="7"} 0.0
# HELP bp_redis_pool_max_size Multiprocess metric
# TYPE bp_redis_pool_max_size gauge
bp_redis_pool_max_size{pid="7",pool="redis"} 0.0
# HELP bp_redis_pool_idle_connections Multiprocess metric
# TYPE bp_redis_pool_idle_connections gauge
bp_redis_pool_idle_connections{pid="7",pool="redis"} 0.0
# HELP bp_redis_pool_active_connections Multiprocess metric
# TYPE bp_redis_pool_active_connections gauge
bp_redis_pool_active_connections{pid="7",pool="redis"} 0.0
# HELP bp_sqlalchemy_pool_max_size Multiprocess metric
# TYPE bp_sqlalchemy_pool_max_size gauge
bp_sqlalchemy_pool_max_size{pid="7",pool="sqlalchemy"} 0.0
# HELP bp_sqlalchemy_pool_idle_connections Multiprocess metric
# TYPE bp_sqlalchemy_pool_idle_connections gauge
bp_sqlalchemy_pool_idle_connections{pid="7",pool="sqlalchemy"} 0.0
# HELP bp_sqlalchemy_pool_active_connections Multiprocess metric
# TYPE bp_sqlalchemy_pool_active_connections gauge
bp_sqlalchemy_pool_active_connections{pid="7",pool="sqlalchemy"} 0.0
# HELP bp_sqlalchemy_pool_overflow_connections Multiprocess metric
# TYPE bp_sqlalchemy_pool_overflow_connections gauge
bp_sqlalchemy_pool_overflow_connections{pid="7",pool="sqlalchemy"} 0.0
# HELP thrift_server_active_requests Multiprocess metric
# TYPE thrift_server_active_requests gauge
thrift_server_active_requests{pid="7",thrift_method="example"} 0.0
thrift_server_active_requests{pid="7",thrift_method="is_healthy"} 0.0
# HELP thrift_client_active_requests Multiprocess metric
# TYPE thrift_client_active_requests gauge
thrift_client_active_requests{pid="7",thrift_method="are_strings_good",thrift_slug="goodcontent"} 0.0
# HELP thrift_client_requests_total Multiprocess metric
# TYPE thrift_client_requests_total counter
thrift_client_requests_total{thrift_baseplate_status="",thrift_baseplate_status_code="",thrift_exception_type="",thrift_slug="goodcontent",thrift_success="true"} 3.0
# HELP thrift_server_requests_total Multiprocess metric
# TYPE thrift_server_requests_total counter
thrift_server_requests_total{thrift_baseplate_status="",thrift_baseplate_status_code="",thrift_exception_type="",thrift_method="example",thrift_success="true"} 3.0
thrift_server_requests_total{thrift_baseplate_status="",thrift_baseplate_status_code="",thrift_exception_type="",thrift_method="is_healthy",thrift_success="true"} 82.0
thrift_server_requests_total{thrift_baseplate_status="",thrift_baseplate_status_code="",thrift_exception_type="TTransportException",thrift_method="example",thrift_success="false"} 2.0
thrift_server_requests_total{thrift_baseplate_status="",thrift_baseplate_status_code="",thrift_exception_type="ServerTimeout",thrift_method="example",thrift_success="false"} 2.0

```

## 🧪 Testing Steps / Validation

## ✅ Checks
- [x] CI tests (if present) are passing
- [x] Adheres to code style for repo
- [ ] Contributor License Agreement (CLA) completed if not a Reddit employee

krav committed:
fixes to thrift client metrics

* set protocol on thrift client

fixing observer ignoring data

* ensure span is started before handling request

when an exception was encontered before the span was started and
additional exception was raised by the `_call_thrift_method` exception
handler, as it expected the span to have been started.

this changes the semantics of the latency graph, as it now includes
every retry of a request!

krav committed:
remove setting protocol from test

nsheaps commented on baseplate/clients/thrift.py:
```diff
@@ -167,17 +167,18 @@ def _call_thrift_method(self: Any, *args: Any, **kwargs: Any) -> Any:
+            client = self.client_cls(prot)
+            method = getattr(client, name)
+            span.set_tag("method", method.__name__)
+            span.start()
```
retrying should start a new span, not one span wrapping all the retries. This change would limit the visibility into why a call failed and should have been retried, along with how many retries and how long each one took.

krav replied:
OK, moving the `for` loop up before this initialization of the span. Initialization needs to exist outside of the `try` block since  `span` must be assigned and `span.start()` must be called before `span.finish()`. Otherwise every exception is rethrown, either an UnboundLocalError or another one about a timer not being started. 

Additionally all tags must be set before `start()` is called, otherwise they're seemingly just dropped. It worries me a bit since that means we need to call `pool.connection` and `client_cls` outside the `try` block and I'm not sure how effectful those are.

nsheaps replied:
There are ways to ensure getting a connection and calling client_cls are still wrapped in a try block and allow for calling `finish()` without raising another exception. Some options:

* Modify `Span` to set an instance variable saying it was started and check it before calling finish
* make `.finish()` gracefully close with a warning if the span was not started (timer error)
* manually track that a span was started in the code and check before calling finish
* wrap those parts in their own try/except.


I see as part of this PR, you took them _out_ of the try block, was that because of the exception handling resulting in throwing an error about the timer not being started?

krav replied:
I don't think you would want to retry if either instantiating client_cls or pool.connection fails, those are probably fatal. So I think this is OK.

> I see as part of this PR, you took them out of the try block, was that because of the exception handling resulting in throwing an error about the timer not being started?

Yes, since `start()` starts the timer and `finish()` expects it to.

nsheaps commented on baseplate/clients/thrift.py:
```diff
@@ -213,38 +214,38 @@ def _call_thrift_method(self: Any, *args: Any, **kwargs: Any) -> Any:
+                    span.finish(exc_info=sys.exc_info())
+                    raise
+                except Error as exc:
+                    # a 5xx error is an unexpected exception but not 5xx are
```
What else would trigger this codepath if not a 5xx error? should this end in "raise" if it's not an error?

If it shouldn't raise because it's a 4xx or something of the like, don't forget to `return result`.

krav replied:
It don't want to change the behavior here, it seems it's expected that the user handle any thrift errors. Why we discard the exception information in the span if it's not 5xx I don't know, the comment seems to indicate it's not interesting? (it's a confusing comment)

krav replied:
Seemingly correct when the endpoint is down:

```
✦3 ❯ kubectl exec -it service/reddit-service-alt-text -- ./test-client example
Traceback (most recent call last):
  File "/src/alt_text/alt_text_thrift/AltTextService-remote", line 119, in <module>
    pp.pprint(client.example())
  File "/src/alt_text/alt_text_thrift/AltTextService.py", line 73, in example
    return self.recv_example()
  File "/src/alt_text/alt_text_thrift/AltTextService.py", line 84, in recv_example
    (fname, mtype, rseqid) = iprot.readMessageBegin()
  File "/usr/local/lib/venv/lib/python3.9/site-packages/thrift/protocol/TBinaryProtocol.py", line 134, in readMessageBegin
    sz = self.readI32()
  File "/usr/local/lib/venv/lib/python3.9/site-packages/thrift/protocol/TBinaryProtocol.py", line 217, in readI32
    buff = self.trans.readAll(4)
  File "/usr/local/lib/venv/lib/python3.9/site-packages/thrift/transport/TTransport.py", line 62, in readAll
    chunk = self.read(sz - have)
  File "/usr/local/lib/venv/lib/python3.9/site-packages/thrift/transport/TTransport.py", line 164, in read
    self.__rbuf = BufferIO(self.__trans.read(max(sz, self.__rbuf_size)))
  File "/usr/local/lib/venv/lib/python3.9/site-packages/thrift/transport/TSocket.py", line 166, in read
    raise TTransportException(type=TTransportException.END_OF_FILE,
thrift.transport.TTransport.TTransportException: TSocket read 0 bytes
command terminated with exit code 1
```

nsheaps replied:
Certainly the naming is weird, as it catches `Error` from [ttypes](https://github.com/reddit/baseplate.py/blob/3037ce02ab1875225ebd68ecb43effd8e77e0f2c/baseplate/thrift/ttypes.py#L226), then catches TException, which error extends.

I am a little confused why we have 3 exception blocks and handle them all similarly, and why some exceptions are worth passing to span.finish() and others are not.

I see the `raise` on L236 so I'm not concerned about my original comment anymore.

That said, your test doesn't use the baseplate thrift client, as evidenced by your stack trace.  Looking at the client script you ran, that's running a binary generated by thrift compilation, which would be why baseplate isn't showing in your stack trace.

Xaelias replied:
Apparently, we treat some of these differently because apparently we consider "Expected" exceptions (like hitting a 4xx, which is the client doing something bad (usually at least...) and other "non server" exceptions to not be failures. So they're not raised/handled the same way.

nsheaps commented on baseplate/clients/thrift.py:
```diff
@@ -213,38 +214,38 @@ def _call_thrift_method(self: Any, *args: Any, **kwargs: Any) -> Any:
+                    last_error = str(exc)
+                    if exc.inner is not None:
+                        last_error += f" ({exc.inner})"
+                    continue
```
🔕  this is what makes the retry mechanism happen, right?

krav replied:
Seemingly, this is the only branch that doesn't end in `raise` or `return`.

nsheaps commented on baseplate/clients/thrift.py:
```diff
@@ -167,17 +167,18 @@ def _call_thrift_method(self: Any, *args: Any, **kwargs: Any) -> Any:
-
+        with self.pool.connection() as prot:
+            span = self.server_span.make_child(trace_name)
+            span.set_tag("protocol", "thrift")
```
to reviewers: this is the fix

nsheaps commented on baseplate/clients/thrift.py:
```diff
@@ -213,38 +214,38 @@ def _call_thrift_method(self: Any, *args: Any, **kwargs: Any) -> Any:
-                if 500 <= exc.code < 600:
+                except TTransportException as exc:
+                    # the connection failed for some reason, retry if able
+                    span.finish(exc_info=sys.exc_info())
```
does `span.finish` tag this info on any of the metrics to understand why a call failed and was retried?

krav replied:
Yes, [it sets `success` and `exception_type`](https://github.com/reddit/baseplate.py/blob/022e020eeeaada76efae75599137cf63bb1bee32/baseplate/observers/prometheus.py#L232-L236)

nsheaps reviewed (changes requested):
I think the span shouldn't be started outside of the retries.

krav committed:
one span per try

Traceback (most recent call last):
  File "/src/alt_text/alt_text_thrift/AltTextService-remote", line 119, in <module>
    pp.pprint(client.example())
  File "/src/alt_text/alt_text_thrift/AltTextService.py", line 73, in example
    return self.recv_example()
  File "/src/alt_text/alt_text_thrift/AltTextService.py", line 84, in recv_example
    (fname, mtype, rseqid) = iprot.readMessageBegin()
  File "/usr/local/lib/venv/lib/python3.9/site-packages/thrift/protocol/TBinaryProtocol.py", line 134, in readMessageBegin
    sz = self.readI32()
  File "/usr/local/lib/venv/lib/python3.9/site-packages/thrift/protocol/TBinaryProtocol.py", line 217, in readI32
    buff = self.trans.readAll(4)
  File "/usr/local/lib/venv/lib/python3.9/site-packages/thrift/transport/TTransport.py", line 62, in readAll
    chunk = self.read(sz - have)
  File "/usr/local/lib/venv/lib/python3.9/site-packages/thrift/transport/TTransport.py", line 164, in read
    self.__rbuf = BufferIO(self.__trans.read(max(sz, self.__rbuf_size)))
  File "/usr/local/lib/venv/lib/python3.9/site-packages/thrift/transport/TSocket.py", line 166, in read
    raise TTransportException(type=TTransportException.END_OF_FILE,
thrift.transport.TTransport.TTransportException: TSocket read 0 bytes
command terminated with exit code 1

nsheaps reviewed (changes requested):
Little bit more discussion to be had

nsheaps committed:
Handle span.finish erroring when parent already closed

nsheaps commented on baseplate/clients/thrift.py:
```diff
@@ -213,38 +218,45 @@ def _call_thrift_method(self: Any, *args: Any, **kwargs: Any) -> Any:
+                    raise
+                except:  # noqa: E722
+                    # something unexpected happened
+                    try:
```
@krav please review this part.

krav replied:
Are we worried that something outside BP is catching this error so ParentSpanAlreadyFinishedError is not reported to sentry?

krav replied:
This changes the behaviour, causing expected exceptions to get missed:

Old behavior:
```
☸ dev-3 (ws-kris-thomt) in baseplate.py on  k/prom-sql-client-metrics [$?] via 🐍 v3.9.13 (alt-text)
✦3 ❯ cat test.py
class SuperBadError(Exception):
    pass

class NotSoBadError(Exception):
    pass

try:
    try:
        raise SuperBadError
    except:
        raise
except SuperBadError:
    print("I am caught!")


☸ dev-3 (ws-kris-thomt) in baseplate.py on  k/prom-sql-client-metrics [$?] via 🐍 v3.9.13 (alt-text)
✦3 ❯ python3 test.py
I am caught!
```

New behavior:
```
☸ dev-3 (ws-kris-thomt) in baseplate.py on  k/prom-sql-client-metrics [$?] via 🐍 v3.9.13 (alt-text)
✦3 ❯ cat test.py
class SuperBadError(Exception):
    pass

class NotSoBadError(Exception):
    pass

try:
    try:
        raise SuperBadError
    except:
        try:
            raise NotSoBadError
        except SuperBadError:
            raise
        raise
except SuperBadError:
    print("I am caught")


☸ dev-3 (ws-kris-thomt) in baseplate.py on  k/prom-sql-client-metrics [$?] via 🐍 v3.9.13 (alt-text)
✦3 ❯ python3 test.py
Traceback (most recent call last):
  File "/home/krav/src/alt-text/baseplate.py/test.py", line 9, in <module>
    raise SuperBadError
__main__.SuperBadError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/krav/src/alt-text/baseplate.py/test.py", line 12, in <module>
    raise NotSoBadError
__main__.NotSoBadError
```

krav replied:
It's not needed, since the exception chain `SomethingError<ParentSpanAlreadyFinishedError` is different from `SomethingError`. The user must explicitly catch `ParentSpanAlreadyFinishedError` for python not to error, `catch SomethingError` is not enough.

nsheaps commented on baseplate/clients/thrift.py:
```diff
@@ -213,38 +218,45 @@ def _call_thrift_method(self: Any, *args: Any, **kwargs: Any) -> Any:
+                    except ParentSpanAlreadyFinishedError:
+                        logger.warn("Attempted to finish a span where the parent was already closed")
+                        raise
+                    except:
```
One of the errors was trying to reference `span` before it was defined, I figured this will be sure anything bubbles up, but I added explicit handling for the case that triggered this change so I'd be fine with dropping this block.

krav replied:
You can safely remove this, since not catching it means raising it.

nsheaps committed:
Revert "Handle span.finish erroring when parent already closed"

This reverts commit f9439a86ae56e3b6b8494049cac7accbde91b82c.
